src/main/host/descriptor/socket/inet/udp.rs

```diff
@@ -0,0 +1,892 @@
+                return Ok(RecvmsgReturn {
+                    return_val: 0,
+                    addr: None,
+                    msg_flags: 0,
+                    control_len: 0,
+                });
+            }
+
+            return Err(SyscallError::new_blocked(
+                File::Socket(Socket::Inet(InetSocket::Udp(socket.clone()))),
+                FileState::READABLE,
+                socket_ref.supports_sa_restart(),
+            ));
+        }
+
+        Ok(result?)
+    }
+
+    pub fn ioctl(
+        &mut self,
+        request: u64,
+        arg_ptr: ForeignPtr<()>,
+        mem: &mut MemoryManager,
+    ) -> SyscallResult {
+        match request {
+            // equivalent to SIOCINQ
+            libc::FIONREAD => {
+                let len = self.recv_buffer.len_bytes().try_into().unwrap();
+
+                let arg_ptr = arg_ptr.cast::<libc::c_int>();
+                mem.write(arg_ptr, &len)?;
+
+                Ok(0.into())
+            }
+            // equivalent to SIOCOUTQ
+            libc::TIOCOUTQ => {
+                let len = self.send_buffer.len_bytes().try_into().unwrap();
+
+                let arg_ptr = arg_ptr.cast::<libc::c_int>();
+                mem.write(arg_ptr, &len)?;
+
+                Ok(0.into())
+            }
+            libc::FIONBIO => {
+                panic!("This should have been handled by the ioctl syscall handler");
+            }
+            libc::TCGETS
+            | libc::TCSETS
+            | libc::TCSETSW
+            | libc::TCSETSF
+            | libc::TCGETA
+            | libc::TCSETA
+            | libc::TCSETAW
+            | libc::TCSETAF
+            | libc::TIOCGWINSZ
+            | libc::TIOCSWINSZ => {
+                // not a terminal
+                Err(Errno::ENOTTY.into())
+            }
+            _ => {
+                log::debug!("We do not yet handle ioctl request {request} on tcp sockets");
+                Err(Errno::EINVAL.into())
+            }
+        }
+    }
+
+    pub fn listen(
+        _socket: &Arc<AtomicRefCell<Self>>,
+        _backlog: i32,
+        _net_ns: &NetworkNamespace,
+        _rng: impl rand::Rng,
+        _cb_queue: &mut CallbackQueue,
+    ) -> Result<(), SyscallError> {
+        Err(Errno::EOPNOTSUPP.into())
+    }
+
+    pub fn connect(
+        socket: &Arc<AtomicRefCell<Self>>,
+        peer_addr: &SockaddrStorage,
+        net_ns: &NetworkNamespace,
+        rng: impl rand::Rng,
+        _cb_queue: &mut CallbackQueue,
+    ) -> Result<(), SyscallError> {
+        // if not an inet socket address
+        let Some(peer_addr) = peer_addr.as_inet() else {
+            return Err(Errno::EINVAL.into());
+        };
+
+        let mut peer_addr: std::net::SocketAddrV4 = (*peer_addr).into();
+
+        // https://stackoverflow.com/a/22425796
+        if peer_addr.ip().is_unspecified() {
+            peer_addr.set_ip(std::net::Ipv4Addr::LOCALHOST);
+        }
+
+        // TODO: handle AF_UNSPEC
```

Comment:
stale comment? looks like it's handled above?

Reply:
This means to handle a `AF_UNSPEC` socket address (as opposed to a `AF_INET` socket address). The above code checks if the IP address is unspecified (is an `AF_INET` socket address with IP "0.0.0.0").

I moved this comment up to where we cast it to an `AF_INET` socket.